qlib/FsUtil.qm

```diff
@@ -0,0 +1,755 @@
+
+        This function will only remove a directory (i.e. path can't point to a regular file, symlink, etc.).
+        This doesn't hold true for the path's contents though of course. That will be removed anyway.
+        This function doesn't follow symlinks. As mentioned in description of REMOVE-TREE-ERROR exception, if only some
+        parts of the tree can't be removed (e.g. due to permissions) and if fail_immediately is False, the tree is
+        partially removed anyway and REMOVE-TREE-ERROR is thrown and describes what exactly went wrong for each error.
+    */
+    public sub remove_tree(string path, bool fail_immediately=True) {
+        Dir dir();
+        dir.chdir(path);
+
+        # remove the directory's contents
+        list errors = ();
+        foreach string name in (dir.list()) {
+            string sub_path = join_paths(path, name);
+            try {
+                if (is_dir(sub_path)) {
+                    remove_tree(sub_path);
+                } else {
+                    remove_file(sub_path);
+                }
+            } catch (err) {
+                if (fail_immediately) {
+                    rethrow;
+                }
+
+                if (err.err == "REMOVE-TREE-ERROR") {
+                    errors += err.desc;
+                } else {
+                    errors += err;
+                }
+            }
+        }
+
+        if (errors) {
+            throw "REMOVE-TREE-ERROR", errors;
+        }
+
+        # directory specified by path should be empty now
+        rmdir(path);
+    }
+
+    #! Universal remove function
+    /** @param path path to be removed
+
+        This wrapper function exists for convenience. Path can be a regular file, symlink or a directory,
+        the respective functions for removing files/links or directories will be called.
+    */
+    public sub remove_path(string path) {
+        if (is_dir(path)) {
+            remove_tree(path);
+        } else {
+            remove_file(path);
+        }
+    }
+
+    #! Checks whether two stat hashes point to the same file or not
+    /** @param stat1 first stat hash
+        @param stat2 second stat hash
+        @param ignore_errors flag indicating that errors should be ignored (True by default)
+
+        @return True if the two stat hashes point to the same file, False otherwise.
+
+        @throws STAT-HASH-ERROR If parameters don't both contain stat hashes.
+
+        Device and inode IDs are compared.
+        If a stat hash does not exist and ignore_errors is True, then the function returns False. This is for
+        convenience of the caller who thus may just call a stat function to a path and pass the result to this function
+        without taking care of possible NOTHINGs. In that case the paths can probably be considered not pointing to
+        the same file.
+    */
+    public bool sub same_file_stat(*hash stat1, *hash stat2, bool ignore_errors=True) {
```

Comment:
```
*hash<StatInfo>
```
as params can reduce the if/else check

Reply:
In the end I used this and at least simplified the condition which unfortunately couldn't be removed completely because you get NOTHING instead of an exception while accessing a key in a `NOTHING` `*hash` :-/

```
string sub star_hash(*hash h) {
    return sprintf("%N\n", h.random_key);
}

*hash my_hash = {"random_key": "value"};
printf("%s\n", star_hash(my_hash));  # will print "value" which is OK

my_hash = {};
printf("%s\n", star_hash(my_hash));  # will print NOTHING which os kinda weird

my_hash = NOTHING;
printf("%s\n", star_hash(my_hash));  # will print NOTHING which is weird

printf("%s\n", NOTHING.random_key);  # will print NOTHING which is even weirder (there is at least a warning though)
```